### PR TITLE
pass-secret-service: Add dbus file, assert

### PR DESCRIPTION
**Pull request by cyntheticfox**
### Description

Add the dbus service file in the package folder to XDG_DATA_HOME, as well as adding an assertion to ensure both it and `gnome-keyring` aren't enabled at the same time.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/services/pass-secret-service.nix [hunk 1]
```diff
@@ -16,33 +15,48 @@ in {
...
     };
   };
 
   config = mkIf cfg.enable {
     assertions = [
       (hm.assertions.assertPlatform "services.pass-secret-service" pkgs
         platforms.linux)
+      {
+        assertion = !config.services.gnome-keyring.enable;
+        message = "Only one secrets service per user can be enabled at a time";
```
Comment (ncfavier) [inline modules/services/pass-secret-service.nix:34]: Maybe mention the conflicting modules

File: modules/services/pass-secret-service.nix [hunk 2]
```diff
@@ -16,33 +15,48 @@ in {
     package = mkPackageOption pkgs "pass-secret-service" { };
 
     storePath = mkOption {
-      type = with types; nullOr str;
-      default = null;
```
Comment (ncfavier) [inline modules/services/pass-secret-service.nix:20]: Why the change in type?

File: modules/services/pass-secret-service.nix [hunk 3]
```diff
@@ -16,33 +15,48 @@ in {
...
-      defaultText = "~/.password-store";
+      type = types.str;
+      default = "";
+      defaultText = "$HOME/.password-store";
```
Comment (ncfavier) [inline modules/services/pass-secret-service.nix:20]: `defaultText` is for reflecting the option's `default` value. If the software uses its own default when the option is `null` or empty, this should be documented in the `description`, or we should just `default` to it to avoid the indirection.

The actual default is also determined by the [password-store module](https://github.com/nix-community/home-manager/blob/669669fcb403e3137dfe599bbcc26e60502c3543/modules/programs/password-store.nix#L62-L63), so maybe this would be best explained with a `literalDocBook`.

File: modules/services/pass-secret-service.nix (hunk 1 above)
Comment (cyntheticfox) [reply modules/services/pass-secret-service.nix:34]: Yeah, I forgot to. I went and double-checked there weren't other services with modules supporting `org.freedesktop.secrets`, but forgot to go back and list them. I'll do so.

File: modules/services/pass-secret-service.nix (hunk 3 above)
Comment (cyntheticfox) [reply modules/services/pass-secret-service.nix:20]: Yeah, I never could decide how best to set the default. I want it to be simple where if they don't define it, it uses the application default, but doesn't forcibly couple itself to the `password-store` module if the person doesn't want to either use `password-store` or wants to have a separate store for the `secrets` service. For example, I don't actually use `password-store` currently outside of the secrets service.

Additionally, while `pass` itself uses the `PASSWORD_STORE_DIR` environment variable, `pass-secret-service` will not respect it, as the underlying project only checks the environment for that variable if you use the CLI (see <https://github.com/aviau/python-pass/blob/7b633e0034ad409872d45f09e965ffbe9ebb9700/pypass/passwordstore.py#L61>), and instead defaults to `$HOME/.password-store`. Previous behavior just configured it with the `password-store` module, which is why we added the `mkDefault` choice.

At the same time, it doesn't make sense to define a `null` choice separately from the empty string `""` IMO, since both would indicate "no path given", and the latter might actually cause things to break.

See previous discussion in <https://github.com/nix-community/home-manager/pull/2969#discussion_r877575663>.

File: modules/services/pass-secret-service.nix (hunk 2 above)
Comment (cyntheticfox) [reply modules/services/pass-secret-service.nix:20]: See below comment

File: modules/services/pass-secret-service.nix (hunk 3 above)
Comment (ncfavier) [reply modules/services/pass-secret-service.nix:20]: > At the same time, it doesn't make sense to define a `null` choice separately from the empty string `""` IMO, since both would indicate "no path given", and the latter might actually cause things to break.

What do you mean? `null` is the standard "no value given" value. The fact that the empty string doesn't make sense as a path (hell, sometimes it does) is accidental.

Comment (ncfavier) [reply modules/services/pass-secret-service.nix:20]: > but doesn't forcibly couple itself to the password-store module if the person doesn't want to either use password-store or wants to have a separate store for the secrets service.

It's a default value. Most people will use a single password store, so it's a good default. Those who don't can change it.

File: modules/services/pass-secret-service.nix (hunk 1 above)
Comment (cyntheticfox) [reply modules/services/pass-secret-service.nix]: Added in b1bfb4dd

File: modules/services/pass-secret-service.nix (hunk 3 above)
Comment (cyntheticfox) [reply modules/services/pass-secret-service.nix:20]: > `null` is the standard "no value given" value.

Really? There are quite a few other modules that contradict that. A good many modules will use an empty list `[ ]`, an empty attrset `{ }` or the empty string `""` rather than `null`, including ones like `kakoune`, `gpg`, and `bspwm`. Heck, Nix even has an entire challenge around attribute paths that may or may not exist, but those don't resolve to `null` when accessed, they throw an error. I'd argue since `""` likely doesn't work due to the provision of a path flag in this context, `""` makes for an adequate "undefined" value, while making the system simpler.

Comment (cyntheticfox) [reply modules/services/pass-secret-service.nix:20]: > > but doesn't forcibly couple itself to the password-store module if the person doesn't want to either use password-store or wants to have a separate store for the secrets service.
> 
> It's a default value. Most people will use a single password store, so it's a good default. Those who don't can change it.

So here's another challenge if we do that: you now _have_ to enable the separate password store, in order for the default configuration to be evaluated. I don't use `password-store` for literally anything that isn't the secrets-service, and while I don't consider myself "most people", it seems awfully bold to presume what most people will do in any case. In any case, that's at best a reason to leave the defaults for both the same value separately, not to couple them together -- and "those who don't can change it"

Comment (ncfavier) [reply modules/services/pass-secret-service.nix:20]: > Really?

In Nix, yes. In languages with a better type system, you would express this type as `Maybe String` or `Option String`, and the default value would be `Nothing` or `None`. `null` is a cheap approximation of that.

In types that have merging semantics, like `lines`, `listOf t` or `attrsOf t`, it makes sense for the default value to be the [neutral element](https://en.wikipedia.org/wiki/Identity_element) of the merging operation (`""`, `[]` and `{}` respectively). This isn't the case here since we can't combine two different paths.

Using `""` as a "default" or "unset" string value is common in shells, for lack of an alternative: everything is a string. Nix has at least an ersatz of a decent type system, let's use it.

Comment (ncfavier) [reply modules/services/pass-secret-service.nix:20]: > do that

Wait, do what? I'm not sure what we're talking about; I'm not suggesting any changes to the modules, only to the documentation.

Comment (ncfavier) [reply modules/services/pass-secret-service.nix:20]: Just revert the type as it was.

File: modules/services/pass-secret-service.nix (hunk 2 above)
Comment (cyntheticfox) [reply modules/services/pass-secret-service.nix]: Reverted in [93d20b1](https://github.com/nix-community/home-manager/pull/3953/commits/93d20b13a018f3fdb5c8e37102b09552c31432b1)

File: modules/services/pass-secret-service.nix (hunk 3 above)
Comment (cyntheticfox) [reply modules/services/pass-secret-service.nix:26]: Should be good with 851df4f, but still waiting on CI to make sure

**Comment by ncfavier** [comment]
The assertion in pass-secret-service is broken in at least two ways. Keep it simple please, this will do:

```nix
      {
        assertion = !config.services.gnome-keyring.enable;
        message = ''
          Only one secrets service per user can be enabled at a time.
          Other services enabled:
          - gnome-keyring
        '';
      }
```

**Review by ncfavier — APPROVED** [review]

File: modules/services/gnome-keyring.nix [hunk 4]
```diff
@@ -28,6 +28,14 @@ in {
     assertions = [
       (lib.hm.assertions.assertPlatform "services.gnome-keyring" pkgs
         lib.platforms.linux)
+      {
+        assertion = !config.services.pass-secret-store.enable;
```
Comment (matejc) [inline modules/services/gnome-keyring.nix:32]: pass-secret-store does not exist, should be pass-secret-service?

Comment (arjan-s) [reply modules/services/gnome-keyring.nix:32]: Looks like it. This merged PR causes the following error on master currently:
```
building the system configuration...
error: attribute 'pass-secret-store' missing

       at /nix/store/bn8h6s75bca34kilcmxgx9kpwz9d4xji-source/modules/services/gnome-keyring.nix:32:22:

           31|       {
           32|         assertion = !config.services.pass-secret-store.enable;
             |                      ^
           33|         message = ''
(use '--show-trace' to show detailed location information)
```

Comment (ncfavier) [reply modules/services/gnome-keyring.nix:32]: *grumble* how did the tests not catch this

Comment (ncfavier) [reply modules/services/gnome-keyring.nix:32]: Oh we don't have a test for gnome-keyring. That would be why.

File: .github/CODEOWNERS [hunk 5]
```diff
@@ -489,7 +489,8 @@ Makefile                                              @thiagokokada
 
 /modules/services/parcellite.nix                      @gleber
 
-/modules/services/pass-secret-service.nix             @cab404
+/modules/services/pass-secret-service.nix             @cab404 @cyntheticfox
+/tests/modules/services/pass-secret-service.nix       @cyntheticfox
```
Comment (rycee) [inline .github/CODEOWNERS:493]: What file is this?

``` console
$ ls tests/modules/services/pass-secret-service.nix
No such file or directory
```

Comment (rycee) [reply .github/CODEOWNERS:493]: Maybe you meant

    /tests/modules/services/pass-secret-service       @cyntheticfox
